### PR TITLE
[Merged by Bors] - Increment last event count on next instead of iter

### PR DESCRIPTION
**Pull request by TheRawMeatball**
# Objective

Currently, simply calling `iter` on an event reader will mark all of it's events as read, even if the returned iterator is never used

## Solution

With this, the cursor will simply move to the last unread, but available event when iter is called, and incremented by one per `next` call.

### CODE REVIEW

**Comment by NathanSWard** [comment]
I like this change quite a bit :)
Though I'm a little worried about people relying on the previous behavior and then this breaking any code that relied on that behavior. 

Also, it would be nice to see a few tests that check this implementation :smile:

**Comment by mockersf** [comment]
this makes CI fails as it keeps the borrow alive for longer, blocking some use of the event reader

**Review by NathanSWard — APPROVED** [review]
This change set looks good to me.
I'm still a bit concerned about people relying on the previous behavior, however, I have no evidence, so I'm ok with the change as is. :)

**Review by alice-i-cecile — APPROVED** [review]
I agree with these changes: this is clearly the correct behavior and this quirk had surprised me before.

**Comment by cart** [comment]
@TheRawMeatball: I also couldn't decouple the lifetimes, so I think it makes sense to just "join" the lifetimes for simplicity, as the current "inheritance" approach is effectively doing that anyway.

File: crates/bevy_ecs/src/event.rs [hunk 1]
```diff
@@ -219,37 +219,17 @@ fn internal_event_reader<'a, T>(
...
+    let b = events.events_b.get(b_index..).unwrap_or_else(|| &[]);
+    let unread_count = a.len() + b.len();
+    *last_event_count = events.event_count - unread_count;
+    let iterator = match events.state {
+        State::A => b.iter().chain(a.iter()),
+        State::B => a.iter().chain(b.iter()),
+    };
+    iterator
+        .map(map_instance_event_with_id)
+        .inspect(move |(_, id)| *last_event_count = (id.id + 1).max(*last_event_count))
```
[the remark this answers is not in the document]
Comment (mockersf) [reply crates/bevy_ecs/src/event.rs:232]: if calling `next_back`, all events in the iterator will be considered consumed if called from a new iterator, but they will still be available in the current iterator.

I think this is the most logical way to do it 👍

**Pull request closed without merging**